### PR TITLE
Fix bug 60848: Incorrect unicode custom attribute decoding

### PR DESCRIPTION
**Pull request by kg**
https://bugzilla.xamarin.com/show_bug.cgi?id=60848

Some possible C# string literals are valid UCS-2 but invalid UTF-16. If one of these literals is passed as an argument to a custom attribute, the resulting assembly will contain invalid UTF-8 strings (because the literal was invalid UTF-16). Windows .NET Framework and .NET Core both handle this by decoding any valid portions of the string as-is, and replacing invalid portions of the string with the unicode replacement character. Presently, we just give up and abort our attempt to decode the string.

This PR updates our string decoding logic for custom attributes to try and match the expected behavior. As far as I can tell, the encoding being used here is effectively WTF-8 (https://simonsapin.github.io/wtf-8/) and it's not meaningful to apply UTF-8 decoding or encoding rules to custom attribute strings. The custom attribute string may not have originally been UTF-8, or even unicode text at all, so we can't make any assumptions when trying to decode it.

After this fix there still remain cases where a custom attribute string will silently fail to decode (and become null), but those cases should be reduced to ones where the data is not even valid if treated as WTF-8 (truncated sequences, etc).

### CODE REVIEW

**Comment by luhenry** [comment]
@monojenkins build Linux i386

**Comment by luhenry** [comment]
@monojenkins build Linux i386 Coop GC

File: mono/eglib/giconv.c [hunk 1]
```diff
@@ -833,15 +833,19 @@ g_utf8_to_ucs4_fast (const gchar *str, glong len, glong *items_written)
 }
 
 static gunichar2 *
-eg_utf8_to_utf16_general (const gchar *str, glong len, glong *items_read, glong *items_written, gboolean include_nuls, GError **err)
-{
+eg_utf8_to_utf16_general (
```
Comment (kumpera) [inline mono/eglib/giconv.c]: Our coding convention calls for all args in the same line.

File: mono/eglib/giconv.c [hunk 2]
```diff
@@ -833,15 +833,19 @@ g_utf8_to_ucs4_fast (const gchar *str, glong len, glong *items_written)
...
+	GError **err
+) {
 	gunichar2 *outbuf, *outptr;
 	size_t outlen = 0;
 	size_t inleft;
 	char *inptr;
 	gunichar c;
 	int u, n;
-	
+
```
Comment (kumpera) [inline mono/eglib/giconv.c]: line damage

File: mono/metadata/custom-attrs.c [hunk 3]
```diff
@@ -213,6 +213,30 @@ load_cattr_enum_type (MonoImage *image, const char *p, const char *boundp, const
 	return mono_class_from_mono_type (t);
 }
 
+static MonoString* 
+cattr_string_from_wtf8 (MonoDomain *domain, const char *text, guint length, MonoError *error)
```
Comment (kumpera) [inline mono/metadata/custom-attrs.c]: This is a generic string conversion, it should go into object.c

Comment (kumpera) [reply mono/metadata/custom-attrs.c]: It should be mono_string_new_wtf8_checked

Comment (jaykrell) [reply mono/metadata/custom-attrs.c]: I thought only "checked" to discriminate from non-checked.

Comment (jaykrell) [reply mono/metadata/custom-attrs.c]: Why does the data have to become utf8 in the first place? We don't store utf16 in the PE and keep it that way?

Comment (kg) [reply mono/metadata/custom-attrs.c]: Custom attribute strings are stored as wtf-8 in .NET PE (the spec says utf-8 but .NET framework and coreclr use wtf-8), but our internal string representation is UTF16 so this conversion is necessary at load time

Comment (jaykrell) [reply mono/metadata/custom-attrs.c]: I guess they were really optimizing the PE for size given typical English names, but imho this is the wrong level to have done that. Oh well.

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by luhenry** [comment]
@monojenkins build Linux ARMv7 hard float interpreter

**Comment by marek-safar** [comment]
@kumpera are you ok with the changes now?

**Pull request merged**